### PR TITLE
Continue implementation of Cap Cons

### PR DESCRIPTION
**Pull request by turbolent**
Work towards #2091 

## Description

- Separate the capability value types into two: path capability and ID capability, i.e. split `StorageCapabilityValue` into `PathCapabilityValue` and `IDCapabilityValue`
- Improve capability controller borrow type fields, make them reference static types
- Implement ID capability borrow/check

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/stdlib/account.go [hunk 1]
```diff
@@ -2989,6 +3009,151 @@ func newPublicAccountCapabilitiesValue(
...
+
+	controllerBorrowType, ok :=
+		inter.MustConvertStaticToSemaType(controllerBorrowStaticType).(*sema.ReferenceType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	if !sema.IsSubType(controllerBorrowType, wantedBorrowType) {
+		return nil, nil
+	}
```
Comment (dsainati1) [inline runtime/stdlib/account.go:3057]: What is the relationship between the type of a controller's borrow type and its controlled capability's borrow type. Do we guarantee that one is a subtype of the other? If so it should be possible to eliminate one of these two subtyping checks.

**Comment by turbolent** [comment]
Was able to address https://github.com/onflow/cadence/pull/2472#issuecomment-1550223638

File: runtime/stdlib/account.go (hunk 1 above)
Comment (turbolent) [reply runtime/stdlib/account.go:3055]: The situation is unfortunately complicated and mainly stems from the behaviour of the `Capability` type/value.

A capability value may have itself a type, which determines how the capability should be borrowed. This is `capabilityBorrowType`. 
Optionally, the capability may then be borrowed with a different type, via `borrow<T>`. This is `wantedBorrowType`.
We first need to ensure that `wantedBorrowType` is not a subtype of `capabilityBorrowType`: https://github.com/onflow/cadence/pull/2476/files#diff-226624e4ec99d61a29cefaff93951c94ce45969551618dc912421100bf851c7bR3025.
In this check, the controller is not yet involved.

Then, here we have to check that the wanted borrow type is also not a subtype of the controller's borrow type, which determines how the controlled capability may get borrowed.

**Review by SupunS — APPROVED** [review]

File: runtime/interpreter/encode.go [hunk 2]
```diff
@@ -1035,7 +1083,11 @@ func (v TypeValue) Encode(e *atree.Encoder) error {
 	}
 
 	// Encode type at array index encodedTypeValueTypeFieldKey
-	return EncodeStaticType(e.CBOR, v.Type)
+	if v.Type == nil {
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/interpreter/encode.go:1086]: Yes, when the `Type` value represents an "unknown"/"invalid" type. Added a comment to the field in 8d778c0d078648d1ea5d27ddb9acf22abe920d4c

File: runtime/interpreter/encode.go [hunk 3]
```diff
@@ -1188,7 +1232,12 @@ func (t OptionalStaticType) Encode(e *cbor.StreamEncoder) error {
 	if err != nil {
 		return err
 	}
-	return EncodeStaticType(e, t.Type)
+
+	if t.Type == nil {
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/interpreter/encode.go]: No it cannot, good catch! Improved in f92fc7b72f55fd4c4da7c9066095ae4c55d04845

**Review by SupunS — APPROVED** [review]
LGTM!

**Pull request merged**